### DISCUSSION TITLE
Load Image with  Content-Type "application/octet-stream”

### DISCUSSION THREAD

**Discussion by DiSleeper** [coil-kt/coil#1124] [Q&A]
Is it possible to load an image with  Content-Type "application/octet-stream”, cause the backend sends a file with this type, Glide can load this content type, but no luck with Coil.
Have an error "
java.lang.IllegalStateException: BitmapFactory returned a null bitmap. Often this means BitmapFactory could not decode the image data read from the input source (e.g. network, disk, or memory) as it's not encoded as a valid image format."

**Comment by lbj1203** [comment]
I have solved it, there are two ways, the first one uses a custom decoder and the other is a custom fetcher.
decoder: get SourceResult.source.source().readByteArray() data, and then convert it by yourself).
fetcher: Write the request by yourself, and then convert the requested data into a readable image file

**Reply by ToleubekRauan** [reply]
@lbj1203  Could you please share some code of your implementation?

**Reply by lbj1203** [reply]
`class CustomDecoder(
    private val result: SourceResult,
    private val options: Options,
    private val imageLoader: ImageLoader
) : Decoder {
    override suspend fun decode(): DecodeResult? {
        result.mimeType?.let {
            if (it.indexOf("application/octet-stream") > -1) {
                val bytes = result.source.source().readByteArray()
                var resultByte: ByteArray? = XImageLoader.decryptOrigin(
                    bytes,
                    XImageLoader.AES_KEY
                )
                val bim = BitmapFactory.decodeByteArray(resultByte, 0, resultByte!!.size)
                return DecodeResult(
                    drawable = bim.toDrawable(options.context.resources),
                    isSampled = false,
                )
            } else {
                return DecodeResult(
                    drawable = toDrawable(result.source.file().toFile()),
                    isSampled = false
                )
            }
        }
        return DecodeResult(drawable = toDrawable(result.source.file().toFile()), isSampled = false)
    }

    private fun toDrawable(file: File): Drawable {
        var drawable: Drawable? = null
        try {
            val fis = FileInputStream(file)
            val btm = BitmapFactory.decodeStream(fis)
            drawable = BitmapDrawable(options.context.resources, btm)
        } catch (e: java.lang.Exception) {
            return ColorDrawable(Color.BLUE)
        }
        return drawable!!
    }

    class Factory() : Decoder.Factory {
        override fun create(
            result: SourceResult,
            options: Options,
            imageLoader: ImageLoader
        ): Decoder? {
            if (isApplicable(result.source.source())) return null
            return CustomDecoder(result, options, imageLoader)
        }

        private fun isApplicable(source: BufferedSource): Boolean {
            return DecodeUtils.isGif(source) ||
                    DecodeUtils.isAnimatedWebP(source) ||
                    (SDK_INT >= 30 && DecodeUtils.isAnimatedHeif(source))
        }

    }
}` 
According to the framework, the file has been downloaded, and then I directly get the data of the file in the file, and then convert it.
If you are using a custom Fetcher, copy the httpFetcher directly, and then parse the data into an image before saving it to disk

**Reply by ToleubekRauan** [reply]
Thank you, btw, what is the "XImageLoader.decryptOrigin" library?

**Comment by lbj1203** [comment]
还有一种是添加Interceptor,拦截数据然后进行转换,再返回
